### DISCUSSION TITLE
.Net Core version of SnapshotUploader

### DISCUSSION THREAD

**Discussion by almostjulian** [microsoft/ApplicationInsights-SnapshotCollector#9] [Ideas]
We'd like to use SnapshotCollector in .NET Core nanoserver container images... can the uploader process get ported over to core?

**Comment by pharring** [comment]
Thanks for the question, @almostjulian.

There's no great answer here.

- We could port it to .NET Core (2.1, I suspect) but then we'd have the "opposite" problem for .NET Framework users who don't have a .NET Core runtime installed.
- So, should we carry two versions? One for .NET Core and another for .NET Framework?
- What if you didn't install the .NET Core runtime with your app, but instead used "Ready-to-run" or some other packaging mechanism? Where would Snapshot Uploader get its runtime pieces? (Remember, it extracts itself from resources into the "ShadowCopy" folder, so the runtime isn't necessarily alongside the SnapshotCollector assembly.)
- Or, should SnapshotUploader itself be a "self-contained" executable?
- Up until quite recently (the 1.4.0 prerelease), we had different targets for net45, netcoreapp2.0 etc, so, in theory, we could include different SnapshotUploader builds for each target. But in 1.4.0-pre we unified to a single netstandard2.0 target (see #5), so that's no longer possible.
- Perhaps SnapshotUploader should be packaged in its own nuget with different targetframeworks (net45 and netcoreapp2.0) and be made a dependency of the SnapshotCollector nuget.

I think, at this point, my preferred approach is to try and make the SnapshotUploader a "self-contained .NET Core application" -- maybe even .NET 5.0. I think that solves or avoids most of the concerns above. The downside, of course, would be the increase in size -- I won't know how much until I try it. It could also be technically challenging since the SnapshotUploader employs its own "resource unpacking" technique to package all its dependent assemblies into a single executable -- although, that technique might be unnecessary once we have a self-contained "package".

BTW, it's great to hear that you're considering using Snapshot Collector in Nanoserver. Do you have it working already in Server Core?

**Reply by almostjulian** [reply]
No, we're actively trying to avoid huge container images, so we were disappointed when we couldn't get it to work with our nanoserver image.   Just curious,  could SnapshotUploader not be made .NET standard 2.0 to align with the rest of the package?

**Reply by pharring** [reply]
It can't be made .NET Standard 2.0 because it's an executable. We need to pick a runtime (.NET framework or .NET Core).

**Reply by pharring** [reply]
I looked briefly into this, but it's complicated by the fact that the bitness of the uploader must match the bitness of the calling app. So we'd need to carry both x86 and x64 versions of the uploader _and_ runtime, doubling the size of the final nuget. There's really no good answer here.

**Reply by KLuuKer** [reply]
i already had to specify my own temp folders, because of ownership problems when you have multiple different apppools trying to write in the same folder
having to grab a specific subpackage (or just a new one) just for .net core support, wouldn't be a issue 

also you might just want to bring your own runtime and trim it heavily  instead of relying on the version that got installed
